overlord/configstate/configcore/hostname.go

```diff
@@ -0,0 +1,95 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+)
+
+func init() {
+	// add supported configuration of this module
+	supportedConfigurations["core.system.hostname"] = true
+}
+
+// We are conservative here and follow hostname(7). The hostnamectl
+// binary is more liberal but let's err on the side of caution for
+// now.
+var validHostname = regexp.MustCompile(`^[a-z0-9][a-z0-9-]{1,62}$`).MatchString
+
+func validateHostnameSettings(tr config.ConfGetter) error {
+	hostname, err := coreCfg(tr, "system.hostname")
+	if err != nil {
+		return err
+	}
+	if hostname == "" {
+		return nil
+	}
+	if !validHostname(hostname) {
+		return fmt.Errorf("cannot set hostname %q: name not valid", hostname)
+	}
+
+	return nil
+}
+
+func handleHostnameConfiguration(tr config.ConfGetter, opts *fsOnlyContext) error {
+	// TODO: convert to "virtual" configuration nodes once we have support
+	// for this. The current code is not ideal because if one calls
+	// `snap get system system.hostname` the answer can be ""
+	// when not set via snap set.
+	//
+	// It will also override any hostname on the next `snap set` run
+	// that was written not using `snap set system system.hostname`.
+	hostname, err := coreCfg(tr, "system.hostname")
+	if err != nil {
+		return nil
+	}
+	// nothing to do
+	if hostname == "" {
```

Comment:
A small nitpick but think about the consequences. There's a concept of static and dynamic hostname. Static hostname is what the host thinks it is called, even while offline. Dynamic hostname is the name given by a particular network. I wonder if it would be sensible to unset the (static) hostname to only see the dynamic hostname.

Dynamic hostnames are commonly granted by DHCP servers.

Reply:
In addition, the manual page for `hostnamectl` has this paragraph:

>          set-hostname NAME
>            Set the system hostname to NAME. By default, this will alter the pretty, the static, and the transient hostname alike; however, if one or more
>            of --static, --transient, --pretty are used, only the selected hostnames are changed. If the pretty hostname is being set, and static or
>            transient are being set as well, the specified hostname will be simplified in regards to the character set used before the latter are updated.
>            This is done by removing special characters and spaces. This ensures that the pretty and the static hostname are always closely related while
>            still following the validity rules of the specific name. This simplification of the hostname string is not done if only the transient and/or
>            static host names are set, and the pretty host name is left untouched.
> 
>            Pass the empty string "" as the hostname to reset the selected hostnames to their default (usually "localhost").

Reply:
@mvo5 ^ your opinion on this would be good before landing this.

Reply:
I was thinking a bit about this, for now I would simply like to ignore it because we are not yet very good at differentiating nil, "", not set etc.